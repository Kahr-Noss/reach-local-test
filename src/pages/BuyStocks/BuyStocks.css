.buy-stocks-wrapper {
  padding: 20px;
  width: 500px;
  margin: 0 auto;
  display: flex;
  flex-direction: column;
}

.data-container {
  display: flex;
  justify-content: space-between;
}

.empty-line{
  height: 20px;
  text-align: right;
}

.buy-input{
  width: 80px;
}