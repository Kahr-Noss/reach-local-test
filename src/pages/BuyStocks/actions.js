import request from 'request-promise';


export const actionsTypes = {
  SELECT_COMPANY: 'SELECT_COMPANY',
  CHANGE_QUANTITY: 'CHANGE_QUANTITY',
  BUY: 'BUY',
  LOAD_PRICE: 'LOAD_PRICE',
  SHOW_ERROR: 'SHOW_ERROR',
}

function selectCompany(code) {
  return (dispatch) => {
    dispatch({ type: actionsTypes.SELECT_COMPANY, code }); // this reset the values and display them as loading
    return request(`https://financialmodelingprep.com/api/v3/stock/real-time-price/${code}`)
      .then((data) => {
        const parsedData = JSON.parse(data);
        if (parsedData.Error) {
          dispatch({ type: actionsTypes.SHOW_ERROR, category: 'company', msg: 'There was a problem loading the price for this company, please chack that you put a valid company code.' });
        }
        dispatch({ type: actionsTypes.LOAD_PRICE, price: parsedData.price });
      })
      .catch((err) => {
        return { type: actionsTypes.SHOW_ERROR, category: 'company', msg: 'There was a problem loading the price for this company, please chack that you put a valid company code.' };
      });
  };
}

function changeQuantity(quantity) {
  return (dispatch) => {

    const parsedQuantity = Number.parseInt(quantity, 10);
    dispatch({ type: actionsTypes.CHANGE_QUANTITY, quantity });
    if (isNaN(parsedQuantity)) {
      dispatch({ type: actionsTypes.SHOW_ERROR, category: 'quantity', msg: 'Please input a number between 1 and 1000' });
    }
    if (quantity > 1000) {
      dispatch({ type: actionsTypes.SHOW_ERROR, category: 'quantity', msg: 'You can only buy a maximum of 1000 stocks in one time.' });
    }
  };
}

function buy(code, price, quantity) {
  if (price * quantity > 1000000) {
    return { type: actionsTypes.SHOW_ERROR, category: 'price', msg: 'You can only buy a maximum of 1.000.000$ worth of stocks in one time.' };
  }
  return {
    type: actionsTypes.BUY,
    code,
    price,
    quantity,
  };
}

export const actions = {
  selectCompany,
  changeQuantity,
  buy,
}