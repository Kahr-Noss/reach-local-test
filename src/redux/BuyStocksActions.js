import request from 'request-promise';
import moment from 'moment';

import isStockExchangeOpen from '../utils/isStockExchangeOpen';

export const actionsTypes = {
  SELECT_COMPANY: 'SELECT_COMPANY',
  CHANGE_QUANTITY: 'CHANGE_QUANTITY',
  BUY: 'BUY',
  LOAD_PRICE: 'LOAD_PRICE',
  SHOW_ERROR_BUY: 'SHOW_ERROR_BUY',
  EDIT_CART: 'EDIT_CART',
  REMOVE: 'REMOVE',
}

let lastAPICallID = 0; // id to check what the last API code was, and discard the rest

function selectCompany(company) {
  const upperCaseCode = company.toUpperCase();
  return (dispatch) => {
    dispatch({ type: actionsTypes.SELECT_COMPANY, company: upperCaseCode }); // this reset the values and display them as loading
    lastAPICallID++;
    const APICallID = lastAPICallID;
    return request(`https://financialmodelingprep.com/api/v3/stock/real-time-price/${encodeURI(upperCaseCode)}`)
      .then((data) => {
        if (APICallID === lastAPICallID) { // only apply the changes if it was the last api call
          const parsedData = JSON.parse(data);
          if (parsedData.Error) {
            dispatch({ type: actionsTypes.SHOW_ERROR_BUY, category: 'company', msg: 'Company code not found.' });
          } else {
            dispatch({ type: actionsTypes.LOAD_PRICE, price: parsedData.price });
          }
        }
      })
      .catch((err) => {
        return { type: actionsTypes.SHOW_ERROR_BUY, category: 'company', msg: 'API is down, please retry later.' };
      });
  };
}

function changeQuantity(quantity, price) {
  return (dispatch) => {
    dispatch({ type: actionsTypes.CHANGE_QUANTITY, quantity });
    if (!quantity.match(/^[0-9]*$/)) {
      dispatch({ type: actionsTypes.SHOW_ERROR_BUY, category: 'quantity', msg: 'Please input a number between 1 and 1000' });
    } else {
      const parsedQuantity = Number.parseInt(quantity, 10);
      if (quantity > 1000) {
        dispatch({ type: actionsTypes.SHOW_ERROR_BUY, category: 'quantity', msg: 'You can only buy a maximum of 1000 stocks in one time.' });
      }
      if (price * quantity > 1000000) {
        dispatch({ type: actionsTypes.SHOW_ERROR_BUY, category: 'total', msg: 'You can only buy a maximum of 1.000.000$ in one time.' });
      }
    }
  };
}

function buyStocks(company, price, quantity) {
  if (quantity > 1000) {
    return { type: actionsTypes.SHOW_ERROR_BUY, category: 'quantity', msg: 'You can only buy a maximum of 1000 stocks in one time.' };
  }
  if (price * quantity > 1000000) {
    return { type: actionsTypes.SHOW_ERROR_BUY, category: 'total', msg: 'You can only buy a maximum of 1.000.000$ in one time.' };
  }

  // check if the stock exchange is open
  if (isStockExchangeOpen(moment())) {
    return {
      type: actionsTypes.BUY,
      company,
      price,
      quantity,
    };
  }
}

function removeStocks(id) {
  return {
    type: actionsTypes.REMOVE,
    id
  };
}

function editCart(id, quantity, price) {
  // only edit the cart is the values are correct
  if (!isNaN(quantity) && quantity <= 1000 && quantity * price <= 1000000) {
    return {
      type: actionsTypes.EDIT_CART,
      id,
      quantity
    };
  }
}

export const actions = {
  selectCompany,
  changeQuantity,
  buyStocks,
  removeStocks,
  editCart,
}