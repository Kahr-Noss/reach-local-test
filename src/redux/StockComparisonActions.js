import request from 'request-promise';


export const actionsTypes = {
  ADD_COMPANY: 'ADD_COMPANY',
  REMOVE_COMPANY: 'REMOVE_COMPANY',
  CHANGE_DATE: 'CHANGE_DATE',
  LOAD_COMPANY_DATA: 'LOAD_COMPANY_DATA',
  CLEAR_COMPANY_DATA: 'CLEAR_COMPANY_DATA',
  SHOW_ERROR: 'SHOW_ERROR'
}

// adding a new company code to the displayed list
function addCompany(code, startDate, endDate) {
  return (dispatch) => {
    dispatch({
      type: actionsTypes.ADD_COMPANY,
      code
    });
    dispatch(loadCompanyData(code, startDate, endDate));
  };
}

// getting data for from a certain company for the specified period
function loadCompanyData(code, startDate, endDate) {
  return (dispatch) => {
    dispatch({ type: actionsTypes.ADD_COMPANY, code }); // this reset the values and display them as loading
    return request(`https://financialmodelingprep.com/api/v3/historical-price-full/${encodeURI(code)}?from=${startDate.format('YYYY-MM-DD')}&to=${endDate.format('YYYY-MM-DD')}`)
      .then((data) => {
        const parsedData = JSON.parse(data);
        if (parsedData.Error) {
          // if the company is not found
          dispatch({ type: actionsTypes.SHOW_ERROR, category: 'company', msg: `Company code not found, please chack that ${code} is a valid company code.`, code });
        } else {
          // keep only relevant data
          const result = {          
            name: parsedData.symbol,
            status: 'complete',
            values: parsedData.historical.map((day) => ({
              date: day.date,
              close: day.close
            }))
          };
          dispatch({ type: actionsTypes.LOAD_COMPANY_DATA, code, result });
        }
      })
      .catch((err) => {
        // website error
        dispatch({ type: actionsTypes.SHOW_ERROR,  msg: 'API is down, please retry later.' });
      });
  }
}

// remove a company from displayed list
function removeCompany(code) {
  return {
    type: actionsTypes.REMOVE_COMPANY,
    code
  }
}

// updating one of the two dates
function changeDate(key, date) {
  return (dispatch) => {
    dispatch({
      type: actionsTypes.CHANGE_DATE,
      key,
      date
    });
  };
}

export const actions = {
  addCompany,
  removeCompany,
  changeDate,
  loadCompanyData,
}