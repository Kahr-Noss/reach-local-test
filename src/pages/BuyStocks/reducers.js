import moment from 'moment';

import { actionsTypes } from './actions';

const initialState = {
  company: '',
  price: null,
  quantity: 1,
  errorList: [],
}

function stockComparisonReducer(state = initialState, action) {
  switch (action.type) {
    case actionsTypes.SELECT_COMPANY: {
      return {
        ...state,
        company: action.code,
        errorList: state.errorList.filter((err) => err.category !== 'company' && err.category !== 'price')
      }
    }

    case actionsTypes.CHANGE_QUANTITY: {
      return {
        ...state,
        quantity: action.quantity,
        errorList: state.errorList.filter((err) => err.category !== 'quantity' && err.category !== 'price')
      }
    }

    case actionsTypes.LOAD_PRICE: {
      return {
        ...state,
        price: action.price
      };
    }

    case actionsTypes.BUY: {
      return initialState;
    }

    case actionsTypes.SHOW_ERROR: {
      return {
        ...state,
        errorList: [
          ...state.errorList,
          {
            msg: action.msg,
            category: action.category
          }
        ]
      }
    }

    default:
      return state
  }
}

export default stockComparisonReducer;