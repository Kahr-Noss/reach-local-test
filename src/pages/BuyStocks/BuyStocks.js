import React from 'react';
import request from 'request-promise';
import moment from 'moment';
import { connect } from 'react-redux';

import { actions } from '../../redux/BuyStocksActions';
import isStockExchangeOpen from '../../utils/isStockExchangeOpen';

// import './BuyStocks.css';
// import "react-datepicker/dist/react-datepicker.css";


export function BuyStocks(props) {

  const isMarketOpen = isStockExchangeOpen(moment());

  return (
    <div className="buy-stocks-wrapper">
      <div className="data-container">
        <div className='text'>Type the company code you want:</div>
        <input className="buy-input" value={props.company} onChange={(e) => props.onCompanyChange(e.target.value)} />
      </div>
      <div className="empty-line error-msg">{props.errors.company}</div>

      <div className="data-container">
        <div className='text'>Stock Price: </div>
        <div>{props.status === 'complete' ? `${props.price}$` : props.status}</div>
      </div>
      <div className="empty-line" />

      <div className="data-container">
        <div className='text'>Quantity:</div>
        <input className="buy-input" value={props.quantity} onChange={(e) => props.onQuantityChange(e.target.value, props.price)} />
      </div>
      <div className="empty-line error-msg">{props.errors.quantity}</div>

      <div className="data-container">
        <div className='text'>Total: </div>
        {isNaN(props.price * props.quantity) ? '' : `${props.price * props.quantity}$`}
      </div>
      <div className="empty-line error-msg">{props.errors.total}</div>

      <br />
      <button
        className="buy-btn btn"
        onClick={() => props.onBuy(props.company, props.price, props.quantity)}
        disabled={!isMarketOpen || !props.company || !props.price || Object.values(props.errors).some((err) => !!err)}
      >
        BUY
      </button>
      {isMarketOpen ? '' : <div className="italic-txt">Sorry, the Tokyo Stock Exchange is currently closed.<br/>Monday to Friday, 09:00-11:30, 12:30-15:00</div>}
    </div>
  );
}

const mapStateToProps = (state) => ({
  company: state.buyStocks.company,
  quantity: state.buyStocks.quantity,
  price: state.buyStocks.price,
  status: state.buyStocks.status,
  errors: state.buyStocks.errors,
});

const mapDispatchToProps = (dispatch) => ({
  onCompanyChange: (company) => { dispatch(actions.selectCompany(company)) },
  onQuantityChange: (quantity, price) => { dispatch(actions.changeQuantity(quantity, price)) },
  onBuy: (company, price, quantity) => { dispatch(actions.buyStocks(company, price, quantity)) },
});


const ConnectedBuyStocks = connect(
  mapStateToProps,
  mapDispatchToProps
)(BuyStocks);


export default ConnectedBuyStocks;