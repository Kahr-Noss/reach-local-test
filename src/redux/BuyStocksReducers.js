import moment from 'moment';

import { actionsTypes } from './BuyStocksActions';

const initialState = {
  company: '',
  price: 0,
  quantity: 1,
  status: 'complete',
  errors: {
    company: null,
    quantity: null,
    total: null,
  },
  cart: []
}

function stockComparisonReducer(state = initialState, action) {
  switch (action.type) {
    case actionsTypes.SELECT_COMPANY: {
      return {
        ...state,
        company: action.company,
        status: 'Loading...',
        errors: {
          ...state.errors,
          company: null,
          total: null,
        }
      }
    }

    case actionsTypes.CHANGE_QUANTITY: {
      return {
        ...state,
        quantity: action.quantity,
        errors: {
          ...state.errors,
          quantity: null,
          total: null,
        }
      }
    }

    case actionsTypes.LOAD_PRICE: {
      return {
        ...state,
        status: 'complete',
        price: action.price
      };
    }

    case actionsTypes.SHOW_ERROR_BUY: {
      return {
        ...state,
        status: action.category === 'company' ? 'Error' : state.status,
        errors: {
          ...state.errors,
          [action.category]: action.msg,
        }
      }
    }

    case actionsTypes.BUY: {
      return {
        ...initialState,
        cart: [
          ...state.cart,
          {
            id: `${action.company.toLowerCase()}${moment().unix()}`,
            time: moment(),
            company: action.company,
            quantity: action.quantity,
            price: action.price
          }
        ]
      };
    }

    case actionsTypes.EDIT_CART: {
      return {
        ...initialState,
        cart: [...state.cart.map((item) => {
          if (item.id !== action.id) {
            return item;
          }
          return {
            ...item,
            quantity: action.quantity
          };
        })]
      };
    }

    case actionsTypes.REMOVE: {
      return {
        ...initialState,
        cart: [...state.cart.filter((item) => item.id !== action.id)]
      };
    }

    default:
      return state
  }
}

export default stockComparisonReducer;