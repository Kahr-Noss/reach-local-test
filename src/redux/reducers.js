import { combineReducers } from 'redux'
import stockComparison from './StockComparisonReducers';
import buyStocks from './BuyStocksReducers';

// combining the two reducers
export default combineReducers({
  stockComparison,
  buyStocks
})