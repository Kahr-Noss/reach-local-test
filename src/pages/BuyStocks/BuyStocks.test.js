import React from 'react';
import { shallow } from 'enzyme';
import configureStore from 'redux-mock-store';
import { Provider } from "react-redux";
import thunk from 'redux-thunk';

import BuyStocks from './BuyStocks';
import { actions } from '../../redux/BuyStocksActions';

import "../../../setupTests";

const mockStore = configureStore([thunk]);

describe('BuyStocks', () => {
  let store;
  let wrapper;
  const initialState = {
    buyStocks: {
      company: 'GOOGL',
      price: 1000,
      quantity: 50,
      status: 'complete',
      errors: {
        company: null,
        quantity: null,
        total: null,
      },
      cart: []
    }
  };
  beforeEach(() => {
    store = mockStore(initialState, actions);
    wrapper = shallow(<BuyStocks store={store}/>);
  });

  // check that store state is correctly passed
  it('Passes props from store', () => {
    expect(wrapper.props().children.props.company).toBe(initialState.buyStocks.company);
    expect(wrapper.props().children.props.price).toBe(initialState.buyStocks.price);
    expect(wrapper.props().children.props.quantity).toBe(initialState.buyStocks.quantity);
    expect(wrapper.props().children.props.status).toBe(initialState.buyStocks.status);
  });

  // check that an action is applied on the store
  it('Input company', () => {
    const company = 'AMZN';
    wrapper.props().children.props.onCompanyChange(company);
    expect(store.getActions()).toContainEqual(actions.selectCompany(company));
  });
});