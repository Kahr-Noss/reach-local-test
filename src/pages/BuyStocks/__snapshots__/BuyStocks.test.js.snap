exports[`test render error inputs properly 1`] = `
<div
  className="buy-stocks-wrapper">
  <div
    className="data-container">
    <div
      className="text">
      Type the company code you want:
    </div>
    <input
      className="buy-input"
      onChange={[Function]}
      value="GOOGL" />
  </div>
  <div
    className="empty-line error-msg">
    error company
  </div>
  <div
    className="data-container">
    <div
      className="text">
      Stock Price: 
    </div>
    <div>
      1000$
    </div>
  </div>
  <div
    className="empty-line" />
  <div
    className="data-container">
    <div
      className="text">
      Quantity:
    </div>
    <input
      className="buy-input"
      onChange={[Function]}
      value={50} />
  </div>
  <div
    className="empty-line error-msg">
    error quantity
  </div>
  <div
    className="data-container">
    <div
      className="text">
      Total: 
    </div>
    50000$
  </div>
  <div
    className="empty-line error-msg">
    error total
  </div>
  <br />
  <button
    className="buy-btn btn"
    disabled={true}
    onClick={[Function]}>
    BUY
  </button>
  <div
    className="italic-txt">
    Sorry, the Tokyo Stock Exchange is currently closed.
    <br />
    Monday to Friday, 09:00-11:30, 12:30-15:00
  </div>
</div>
`;

exports[`test render standard inputs properly 1`] = `
<div
  className="buy-stocks-wrapper">
  <div
    className="data-container">
    <div
      className="text">
      Type the company code you want:
    </div>
    <input
      className="buy-input"
      onChange={[Function]}
      value="GOOGL" />
  </div>
  <div
    className="empty-line error-msg" />
  <div
    className="data-container">
    <div
      className="text">
      Stock Price: 
    </div>
    <div>
      1000$
    </div>
  </div>
  <div
    className="empty-line" />
  <div
    className="data-container">
    <div
      className="text">
      Quantity:
    </div>
    <input
      className="buy-input"
      onChange={[Function]}
      value={50} />
  </div>
  <div
    className="empty-line error-msg" />
  <div
    className="data-container">
    <div
      className="text">
      Total: 
    </div>
    50000$
  </div>
  <div
    className="empty-line error-msg" />
  <br />
  <button
    className="buy-btn btn"
    disabled={true}
    onClick={[Function]}>
    BUY
  </button>
  <div
    className="italic-txt">
    Sorry, the Tokyo Stock Exchange is currently closed.
    <br />
    Monday to Friday, 09:00-11:30, 12:30-15:00
  </div>
</div>
`;
